File: 2024/python/day_02.py
"""Day 2"""

import itertools


def puzzle_input(file):
    with open(file, "r+") as file:
        return [r.strip().split(" ") for r in file.readlines()]


def piece1(x):
    xi = map(int, x)
    pairs = list(itertools.pairwise(xi))
    greater_than_last = itertools.starmap(lambda x, y: x > y, pairs)
    less_than_last = itertools.starmap(lambda x, y: x < y, pairs)
    difference_between_zero_four = itertools.starmap(
        lambda x, y: 0 < abs(x - y) < 4, pairs
    )
    return (all(greater_than_last) or all(less_than_last)) and all(
        difference_between_zero_four
    )


def piece2(x):
    p1 = piece1(x)
    if p1:
        return True
    for i in range(len(x)):
        xc = x.copy()
        xc.pop(i)
        if piece1(xc):
            return True
    return False


def part1(x):
    """
    >>> part1([
    ...     [8,6,4,2,1],
    ...     [1,2,7,8,9],
    ...     [9,7,6,2,1],
    ...     [1,3,2,4,5],
    ...     [8,6,4,4,1],
    ...     [1,3,6,7,9]
    ... ])
    2
    """
    return sum(map(piece1, x))


def part2(x):
    """
    >>> part2([
    ...     [8,6,4,2,1],
    ...     [1,2,7,8,9],
    ...     [9,7,6,2,1],
    ...     [1,3,2,4,5],
    ...     [8,6,4,4,1],
    ...     [1,3,6,7,9]
    ... ])
    4
    """
    return sum(map(piece2, x))


if __name__ == "__main__":
    x = puzzle_input("inputs/02.txt")
    print("Part 1:", part1(x))
    print("Part 2:", part2(x))

File: 2015/R/day_02.r
#' Day 02

puzzle_input <- function(file) {
  readLines(file) |>
    strsplit(split = "x") |>
    lapply(as.numeric) |>
    lapply(as.list) |>
    lapply(setNames, c("l", "w", "h"))
}

part1 <- function(x) {
  box <- sapply(x, function(x) with(x, 2 * l * w + 2 * w * h + 2 * h * l))
  extra <- sapply(x, function(e) prod(sort(unlist(e))[1:2]))
  sum(box + extra)
}

part2 <- function(x) {
  present <- sapply(x, function(e) sum(sort(unlist(e))[1:2] * 2))
  bow <- sapply(x, function(x) with(x, l * w * h))
  sum(present + bow)
}

x <- puzzle_input("inputs/02.txt")

cat("Part 1:", part1(x), "\n")
cat("Part 2:", part2(x), "\n")
